Bottom: 4e219015cf88dd1203aef9841323948c017a9022
Top:    29211ad99c44ccf88d4377779d0731dc37597dfc
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-01-11 10:31:05 +0800

Refresh of vhost_net-batch-tx-descriptor-0

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 074628d..b7913c9 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -349,6 +349,8 @@ static struct vhost_desc *vhost_net_tx_get_vq_desc(struct vhost_net *net,
 	if (desc && desc->head && desc->head != vq->num) {
 		*out_num = desc->out_num;
 		*in_num = desc->in_num;
+	} else {
+		*out_num = *in_num = 0;
 	}
 
 	return desc;
@@ -404,6 +406,9 @@ static void handle_tx(struct vhost_net *net)
 			break;
 
 		desc = vhost_net_tx_get_vq_desc(net, vq, &out, &in);
+		/* TODO: for safe! */
+		if (desc == NULL)
+			break;
 		head = desc->head;
 		/* On error, stop handling until the next kick. */
 		if (unlikely(head < 0))
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 9e667d1..15e804b 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2068,7 +2068,7 @@ struct vhost_desc *vhost_get_vq_desc_batched(struct vhost_virtqueue *vq,
 	if (vq->max_desc) {
 		ret = &vq->descs[vq->current_desc];
 
-		if (vq->current_desc == vq->max_desc)
+		if (vq->current_desc + 1 == vq->max_desc)
 			vq->current_desc = vq->max_desc = 0;
 		else
 			vq->current_desc++;
@@ -2088,10 +2088,8 @@ struct vhost_desc *vhost_get_vq_desc_batched(struct vhost_virtqueue *vq,
 		if (desc->head || desc->head != vq->num) {
 			desc->offset = offset;
 			offset += desc->out_num + desc->in_num;
-		} else {
-			desc->out_num = desc->in_num = 0;
+		} else
 			break;
-		}
 
 		if (offset > UIO_MAXIOV)
 			break;
